### PR TITLE
feat(AWS IAM): Allow `tags` parameter on lambda execution role

### PR DESCRIPTION
**Pull request by runk**
Provide consumers means for setting additional/custom tags to the lambda execution role.

Closes: #8669

### CODE REVIEW

File: lib/plugins/aws/package/lib/mergeIamTemplates.js
Comment (runk) [inline lib/plugins/aws/package/lib/mergeIamTemplates.js:77]: I noticed at least 5 instances of people doing this kind of stuff across codebase - what's your view on making a small lib for this? Pretty basic code, unsure if it's a worthwhile.

File: test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js
Comment (runk) [inline test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js:344]: Let me know if you're okay with this - I know you were favouring single `runServerless` invocation per test.

File: lib/plugins/aws/package/lib/mergeIamTemplates.js
Comment (pgrzesik) [reply lib/plugins/aws/package/lib/mergeIamTemplates.js:77]: I think that while it might make sense, I think it could be submitted as a separate refactoring PR/commit instead of adding it to this one

File: test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js
Comment (pgrzesik) [reply test/unit/lib/plugins/aws/package/lib/mergeIamTemplates.test.js:344]: That's perfect :+1:

**Review by pgrzesik — CHANGES_REQUESTED** [review]
Thank you @runk :raised_hands: There are a few small things that would be good to add:
- Let's add a node to `iam.md` documentation about `tags` property
- Let's add that field to `serverless.yml.md` file 

After that we should be good to go I believe :100:

**Comment by runk** [comment]
@pgrzesik please check https://github.com/serverless/serverless/pull/9059 first - I figured some docs were out of date, perhaps changing all that in this PR is out of scope

**Comment by pgrzesik** [comment]
Thanks @runk, good call with separate PR - already merged it :bow:

**Comment by runk** [comment]
@pgrzesik I believe I addressed the feedback, please check again

**Comment by pgrzesik** [comment]
Thanks @runk - before merging I'd like to consult @medikoo about one thing when it comes to tags as we had a similar discussion recently related to `apiGateway`. Do you think we should expose `tags` properties on separate entities like in this case on `LambdaExecutionRole` or apply the tags set on `provider` level as in case of `apiGateway`? If we should allow the `tags` property here, what should be the rule as to where we allow such customization and where we apply the ones from `provider.tags`? 

@runk - what is your opinion on the above?

**Comment by runk** [comment]
I see, now there are two properties - `provider.tags` and `provider.stackTags`. To make sure we all on the same page re intended usage:

 - `tags` - apply to all functions
 - `stackTags` - CF stack tags that apply to ALL managed resources

So today `provider.tags` do not apply for the execution role created. Obviously `stackTags` do.

I see value in being specific with tags - e.g. setting some specific tags to the role and not everything else. This is the purpose of my change.

**Comment by medikoo** [comment]
I've investigated history of this properties, and it appears that:

1. First `provider.stackTags` (to resemble similar property that was supported with v0.5) were introduced with https://github.com/serverless/serverless/pull/2158
2. Then `provider.tags` were introduced as global version of `functions[].tags` (so indeed purely with intention to target functions) -> https://github.com/serverless/serverless/pull/5007
3. Then support for configuration API Gateway Stage resource was introduced, and over there it was decided to adapt both `provider.stackTags` and `provider.tags` for it -> https://github.com/serverless/serverless/pull/5851
4. Then (recently) we've added support to expose `provider.tags` on HTTP API

So after all it's quite quirky. Assuming we're capable to reset this to something logical, what configuration would you envision?

Does it make sense to have some global _tags_ property, through which user may configure tags to be set for any resource as created by the Framework which support tags? (e.g. CF stack, function, API Gateway, etc) ?

And then provide more settings for specific groups, that allow to tag e.g. all functions, and then ability to setup tags for specific function (we have it now via `functions[].tags`) ?

**Comment by medikoo** [comment]
@pgrzesik @runk I think in scope of this PR, question is, whether aside of `provider.iam.tags` we should also assign `provider.tags`.

Assuming that we will drive `provider.tags` as a collection of tags to be set on every resource as created by Framework, it'll be nice to do it, still that probably should be behind a flag as we did with HTTP API.

**Comment by pgrzesik** [comment]
> So after all it's quite quirky. Assuming we're capable to reset this to something logical, what configuration would you envision?

I was thinking about it a bit and started wondering - do we want to make `provider.tags` to be applied universally to created resources? My first thought was that for that purpose, we should use `stackTags` and let CloudFormation take care of that. But then, `provider.tags` sit kinda in between, because the fact that they're defined on `provider` level suggests that they will be applied universally, but right now they only apply to `apiGateway` and `functions`.

**Comment by runk** [comment]
Okay added couple of lines of code to merge `provider.tags` with `iam.role.tags`. 🤷‍♂️ 

Agree with @pgrzesik - if we'd universally apply `provider.tags`, it'd behave same way `provider.stackTags` does. The only difference is AWS magically applies tags behind scenes in case of `stackTags` while we're being very explicit (+ added code complexity) with `provider.tags` option. Those two props carry very similar meaning..

**Comment by pgrzesik** [comment]
I think we should hold off with merging `provider.tags` with `iam.role.tags` @runk or make it work differently. The approach of applying them without user's explicit opt-in might be a breaking change in situation where someone already tagged role manually and don't even know about the existence of `iam.role.tags` - in such case, these manual tags will be replaced with the ones set on `provider` level. If we would like to apply them, I think it should be done behind a flag and with a deprecation notice that it will be default moving forward. 

@medikoo What do you think? Should we merge tags from `provider.tags` to `iam.role.tags` ?

**Comment by medikoo** [comment]
I actually missed the fact that `provider.stackTags` end not only CF stack tags, but are also propagated to CF resources as deployed with a stack (that also explains to me why they're merged with `provider.tags` on API Gateway). That changes the perspective.

_I think design issue on our side is that function-wide tags were introduced as  `provider.tags`. It'll be cleaner if it's `provider.function.tags` (same as e.g. we nest API Gateway settings on `provider.apiGateway`)_

Anyway, focusing on scope we have here, now, I totally agree, now, that we should not involve `provider.tags` here.

Still if defining the `provider.iam.role.tags` will suddenly remove from IAM resource tags as eventually defined at `provider.stackTags` I believe that we should combine those two collections for IAM role tags setting (same as it's done in API Gateway case). What do you think?

And otherwise (but that's ofc not in scope here), we may cleanup our tags configuration, so it's not confusing by introducing following changes:
- Rename `provider.tags` to `provider.function.tags`
- Rename `provider.stackTags` to `provider.tags`

Ofc that'll be breaking and will require to be put behind the flag.

**Comment by pgrzesik** [comment]
> Still if defining the provider.iam.role.tags will suddenly remove from IAM resource tags as eventually defined at provider.stackTags I believe that we should combine those two collections for IAM role tags setting (same as it's done in API Gateway case). What do you think?

That is not the case. The `stackTags` are applied differently and they're not overwriten by explicit tags set on individual resources defined within CloudFormation template. Please see the example below - there are tags applied to a Lambda function, the last one is coming from `provider.tags` and the rest are default CF Stack tags plus one default that is applied by the framework as a Stack Tag

tags

So in this particular case, I think we might stick to just supporting tags defined on `iam.role.tags` without merging the ones from `provider.tags`. 

> And otherwise (but that's ofc not in scope here), we may cleanup our tags configuration, so it's not confusing by introducing following changes:
> Rename provider.tags to provider.function.tags
> Rename provider.stackTags to provider.tags
> Ofc that'll be breaking and will require to be put behind the flag.

The above is a bit more tricky I believe - `provider.tags` are currently also applied to `apiGateway` so we would have to highlight that as well. I think we should keep the `provider.stackTags` without migrating to `provider.tags` as it might be confusing. But I think this is a discussion for a totally separate issue :+1:

**Comment by medikoo** [comment]
> The stackTags are applied differently and they're not overwriten by explicit tags set on individual resources defined within CloudFormation template. 

@pgrzesik great thanks for checking that! That's good to hear, in API Gateway their used as there tags are set independently via AWS SDK (not deployed via SDK). Still I was puzzled a bit as initial implemtation which was trying to drive it through CF was also merging two collections.

> The above is a bit more tricky I believe - provider.tags are currently also applied to apiGateway so we would have to highlight that as well. 

Yes, it'll be a global change to how tags are configiured in the service, so it'll affect anything that depends on `provider.tags` and `provider.stackTags`.

>  I think we should keep the provider.stackTags without migrating to provider.tags as it might be confusing. But I think this is a discussion for a totally separate issue 👍

My point was to have global tags setting (which apply to all deployed resources) and global function tags setting (that applies just to functions and resources that derive from function events) meaningfully named. Currently they're named as `stackTags` and `tags`, that doesn't feel intuitive to me. Anyway as you said it's separate issue.

**Comment by runk** [comment]
Okay, I reverted some `provider.tags` related code from my PR, please check again.

**Review by pgrzesik — APPROVED** [review]
Thanks @runk :bow: It looks great :+1: As for the more general discussion on `tags`, it will be moved to a separate issue :+1:

**Pull request merged**
